File: implement-shell-tools/wc/wc.js
#!/home/ec2-user/.nvm/versions/node/v24.19.0/bin/node

import { program } from "commander";
import { promises as fs } from "node:fs";
import process from "node:process";

const TOOL_NAME = process.argv[1].split("/").pop();
const TOOL_AUTHOR = "Wyatt L.";
const TOOL_VERSION = `${TOOL_NAME} (CYF shelltools) 1.00\n\nWritten by ${TOOL_AUTHOR}`;
const TOOL_CAVEAT = ``;

program
    .name(TOOL_NAME)
    .usage("[OPTION]... [FILE]...\n  or:  "+TOOL_NAME+" [OPTION]... --files0-from=F")
    .description("Print newline, word, and byte counts for each FILE, and a total line if\nmore than one FILE is specified.  A word is a non-zero-length sequence of\ncharacters delimited by white space.\n\nWith no FILE, or when FILE is -, read standard input.\n\nThe options below may be used to select which counts are printed, always in\nthe following order: newline, word, character, byte, maximum line length.")
    .option("-c, --bytes", "print the byte counts")
    .option("-l, --lines", "print the newline counts")
    .option("-m, --chars", "print the character counts")
    .option("-w, --words", "print the word counts")
    .helpOption("--help", "display this help and exit")
    .version(TOOL_VERSION, "--version", "output version information and exit")
    .addHelpText("after", TOOL_CAVEAT)
    .argument("[FILE...]", null, ["-"])
    .parse();

const files = program.args;
const isByte = (Object.keys(program.opts()).length==0 ? true : (program.opts().bytes===undefined ? false : program.opts().bytes));
const isChar = (Object.keys(program.opts()).length==0 ? false : (program.opts().chars===undefined ? false : program.opts().chars));
const isLine = (Object.keys(program.opts()).length==0 ? true : (program.opts().lines===undefined ? false : program.opts().lines));
const isWord = (Object.keys(program.opts()).length==0 ? true : (program.opts().words===undefined ? false : program.opts().words));

// commander not correctly specify default value for an optional command-argument
if (files.length==0) {
    files.push("-");
}

let counts = {line:0, word:0, char:0, byte:0};
let totals = {line:0, word:0, char:0, byte:0};
let pad = 7;
let content = "";

// https://dmitripavlutin.com/what-every-javascript-developer-should-know-about-unicode/
async function countTotal(filename, isSum) {
    try {
        content = await fs.readFile(filename, "utf-8");  // TODO: should echo line immediately when pressing enter
        counts.line = (content.match(/\n/g)==null ? 0 : content.match(/\n/g).length);
        counts.word = (content.match(/\S+/g)==null ? 0 : content.match(/\S+/g).length);
        counts.char = [...content].length;
        counts.byte = Buffer.byteLength(content);
        if (isSum) {
            totals.line += counts.line;
            totals.word += counts.word;
            totals.char += counts.char;
            totals.byte += counts.byte;
        }
    }
    catch (error) {
        throw error;
    }
}

for (let file of files) {
    if (file!="-") {
        try {
            await countTotal(file, true);
        }
        catch {
            //console.error(error.message.split(",")[0].replace(/^[A-Z]*:/, `${TOOL_NAME}: ${file}:`));
        }
    }
}
if (!files.includes("-")) {
    pad = 0;
    if (isLine && totals.line.toString().length>pad) {
        pad = totals.line.toString().length;
    }
    if (isWord && totals.word.toString().length>pad) {
        pad = totals.word.toString().length;
    }
    if (isChar && totals.char.toString().length>pad) {
        pad = totals.char.toString().length;
    }
    if (isByte && totals.byte.toString().length>pad) {
        pad = totals.byte.toString().length;
    }
}

for (let file of files) {
    try {
        await countTotal((file=="-" ? "/dev/stdin" : file), false);
console.log(counts, file);
    }
    catch (error) {
        console.error(error.message.split(",")[0].replace(/^[A-Z]*:/, `${TOOL_NAME}: ${file}:`));
    }
}
console.log(totals, "total");

File: implement-shell-tools/wc/wc.py
#!/usr/bin/python3

import re
import argparse
import sys

TOOL_NAME = sys.argv[0].rsplit("/", 1)[1]
TOOL_AUTHOR = "Wyatt L."
TOOL_VERSION = "%s (CYF shelltools) 1.00\n\nWritten by %s" % (TOOL_NAME, TOOL_AUTHOR)
TOOL_CAVEAT = ""

parser = argparse.ArgumentParser(
    prog=TOOL_NAME,
    usage="%s [OPTION]... [FILE]...\n  or:  %s [OPTION]... --files0-from=F" % (TOOL_NAME, TOOL_NAME),
    description='''
Print newline, word, and byte counts for each FILE, and a total line if
more than one FILE is specified.  A word is a non-zero-length sequence of
characters delimited by white space.

With no FILE, or when FILE is -, read standard input.

The options below may be used to select which counts are printed, always in
the following order: newline, word, character, byte, maximum line length.
''',
    epilog=TOOL_CAVEAT,
    add_help=False,
    formatter_class=argparse.RawDescriptionHelpFormatter
)
parser.add_argument("-c", "--bytes", action="store_true", help="print the byte counts")
parser.add_argument("-l", "--lines", action="store_true", help="print the newline counts")
parser.add_argument("-m", "--chars", action="store_true", help="print the character counts")
parser.add_argument("-w", "--words", action="store_true", help="print the word counts")
parser.add_argument("--help", action="help", help="display this help and exit")
parser.add_argument("--version", action="version", help="output version information and exit", version=TOOL_VERSION)
parser.add_argument("FILE", nargs="*", default=[""])
args = parser.parse_args()

files = args.FILE
is_byte = True if args.bytes==False and args.chars==False and args.words==False and args.lines==False else args.bytes
is_char = False if args.bytes==False and args.chars==False and args.words==False and args.lines==False else args.chars
is_word = True if args.bytes==False and args.chars==False and args.words==False and args.lines==False else args.words
is_line = True if args.bytes==False and args.chars==False and args.words==False and args.lines==False else args.lines

counts = {"line":0, "word":0, "char":0, "byte":0}
totals = {"line":0, "word":0, "char":0, "byte":0}
pad_size = 7  # default padding size when any file is stdin
content = ""

def count(filename):
    try:
        with open(filename, "r") as f:
            content = f.read()
            counts["line"] = content.count("\n")
            counts["word"] = len(re.findall("\S+", content))
            counts["char"] = len(content)
            counts["byte"] = len(content.encode("utf-8"))
    except Exception as error:
        raise

# prepare padding size when no file is stdin
if ("-" not in files) and ("" not in files):
    for file in files:
        try:
            count(file)
            totals = {key: totals[key]+counts[key] for key in totals}  # totals[____] += counts[____]
        except Exception as error:
            pass #print(re.sub("\[.*?\]", TOOL_NAME+": "+file+":", str(error).split(":")[0]))
    pad_size = 0
    if is_line and (len(str(totals["line"]))>pad_size):
        pad_size = len(str(totals["line"]))
    if is_word and (len(str(totals["word"]))>pad_size):
        pad_size = len(str(totals["word"]))
    if is_char and (len(str(totals["char"]))>pad_size):
        pad_size = len(str(totals["char"]))
    if is_byte and (len(str(totals["byte"]))>pad_size):
        pad_size = len(str(totals["byte"]))
    totals = {key: 0 for key in totals}  # totals[____] = 0

# count and print each file including stdin
for file in files:
    try:
        count(file if (file!="-" and file!="") else "/dev/stdin")
        totals = {key: totals[key]+counts[key] for key in totals}  # totals[____] += counts[____]
        if is_line:
            print(str(counts["line"]).rjust(pad_size, " ")+ " ", end="")
        if is_word:
            print(str(counts["word"]).rjust(pad_size, " ")+ " ", end="")
        if is_char:
            print(str(counts["char"]).rjust(pad_size, " ")+ " ", end="")
        if is_byte:
            print(str(counts["byte"]).rjust(pad_size, " ")+ " ", end="")
        print(file)
    except Exception as error:
        print(re.sub("\[.*?\]", TOOL_NAME+": "+file+":", str(error).split(":")[0]))
if len(files)>1:
    if is_line:
        print(str(totals["line"]).rjust(pad_size, " ")+ " ", end="")
    if is_word:
        print(str(totals["word"]).rjust(pad_size, " ")+ " ", end="")
    if is_char:
        print(str(totals["char"]).rjust(pad_size, " ")+ " ", end="")
    if is_byte:
        print(str(totals["byte"]).rjust(pad_size, " ")+ " ", end="")
    print("total")